src/main/java/ch/njol/skript/util/ListVariablePersistentDataType.java

```diff
@@ -62,32 +65,34 @@
 
 	@Override
 	public byte[] toPrimitive(Map<String, Value> complex, PersistentDataAdapterContext context) {
-		int bufferLength = 0;
-
-		for (Entry<String, Value> entry : complex.entrySet()) {
-			// Store it: index -> type -> data
-			bufferLength += INT_LENGTH + entry.getKey().getBytes(SERIALIZED_CHARSET).length
-						+ INT_LENGTH + entry.getValue().type.getBytes(SERIALIZED_CHARSET).length 
-						+ INT_LENGTH + entry.getValue().data.length;
-		}
-
-		ByteBuffer bb = ByteBuffer.allocate(bufferLength);
+		ByteBuffer intConverter = ByteBuffer.allocate(4);
+		ByteArrayOutputStream byteArray = new ByteArrayOutputStream();
 
 		for (Entry<String, Value> entry : complex.entrySet()) {
 			byte[] indexBytes = entry.getKey().getBytes(SERIALIZED_CHARSET);
 			byte[] typeBytes = entry.getValue().type.getBytes(SERIALIZED_CHARSET);
 
-			bb.putInt(indexBytes.length);
-			bb.put(indexBytes);
-
-			bb.putInt(typeBytes.length);
-			bb.put(typeBytes);
-
-			bb.putInt(entry.getValue().data.length);
-			bb.put(entry.getValue().data);
+			try {
```

Comment:
could you explain why the change from ByteBuffer to ByteArrayOutputSTream was required?

Reply:
It wasn't required, but it'd be more convenient than having to pre-allocate the size of the buffer by calculating the length in bytes.
Also it may be better to use `DataOutputStream` as well to avoid having to use ByteBuffer just for the int writing

Reply:
Sure, I'll take a look at making further improvements here. I would also like to add that I found better performance with this new method compared to the old method where it was manually calculated.